File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/PhoneInviteTest.kt
package org.wyrdsekai.app.engine

import org.wyrdsekai.app.engine.discovery.PhoneInvite
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * wyrdphone:// invite parsing. Fixtures mirror
 * what registration.py mint_phone_invite emits (compact JSON, base64url,
 * no padding).
 */
@OptIn(ExperimentalEncodingApi::class)
class PhoneInviteTest {

    private fun encode(host: String, json: String): String {
        val b64 = Base64.UrlSafe.withPadding(Base64.PaddingOption.ABSENT)
            .encode(json.encodeToByteArray())
        return "wyrdphone://$host/$b64"
    }

    private val selfSignedJson = """
        {"household_id":"hh-9ce1d3b57ebf","kind":"phone","minted_at":1781225931,
         "relays":[{"ca_fp":"E5:F0:A2:3E","fp":"63:24:CB:6B",
                    "nats_password":"pw-secret","nats_user":"relay_phone",
                    "ws_url":"wss://127.0.0.1:4443"}],
         "v":1,"zone_id":"unspecified"}
    """.trimIndent()

    @Test
    fun detectsInviteUrls() {
        assertTrue(PhoneInvite.isPhoneInviteUrl("  wyrdphone://x/abc "))
        assertTrue(PhoneInvite.isPhoneInviteUrl("WYRDPHONE://x/abc"))
        assertFalse(PhoneInvite.isPhoneInviteUrl("wyrdrelay://x/abc"))
        assertFalse(PhoneInvite.isPhoneInviteUrl("https://example.org"))
    }

    @Test
    fun parsesSelfSignedInvite() {
        val invite = PhoneInvite.parse(encode("127.0.0.1:4443", selfSignedJson))
        assertEquals(1, invite.relays.size)
        val r = invite.relays[0]
        assertEquals("wss://127.0.0.1:4443", r.wsUrl)
        assertEquals("relay_phone", r.natsUser)
        assertEquals("pw-secret", r.natsPassword)
        assertEquals("63:24:CB:6B", r.fp)
        assertEquals("E5:F0:A2:3E", r.caFp)
        assertEquals("hh-9ce1d3b57ebf", invite.householdId)
        assertEquals(1781225931L, invite.mintedAt)
        // "unspecified" sentinel maps to null.
        assertNull(invite.zoneId)
    }

    @Test
    fun parsesAcmeInviteWithoutPinMaterial() {
        val invite = PhoneInvite.parse(encode("relay.example.org",
            """{"kind":"phone","relays":[{"nats_password":"pw",
                "nats_user":"relay_phone","ws_url":"wss://relay.example.org"}],"v":1}"""))
        assertNull(invite.relays[0].fp)
        assertNull(invite.relays[0].caFp)
    }

    @Test
    fun preservesFailoverOrdering() {
        val invite = PhoneInvite.parse(encode("a",
            """{"kind":"phone","relays":[
                {"nats_password":"p1","nats_user":"u1","ws_url":"wss://first"},
                {"nats_password":"p2","nats_user":"u2","ws_url":"wss://second"}],"v":1}"""))
        assertEquals(listOf("wss://first", "wss://second"), invite.relays.map { it.wsUrl })
    }

    @Test
    fun rejectsMalformedInput() {
        assertFailsWith<IllegalArgumentException> { PhoneInvite.parse("https://nope") }
        assertFailsWith<IllegalArgumentException> { PhoneInvite.parse("wyrdphone://host-only") }
        assertFailsWith<IllegalArgumentException> { PhoneInvite.parse("wyrdphone://h/!!notb64!!") }
        assertFailsWith<IllegalArgumentException> {
            PhoneInvite.parse(encode("h", """{"kind":"zone","relays":[],"v":1}"""))
        }
        assertFailsWith<IllegalArgumentException> {
            PhoneInvite.parse(encode("h", """{"kind":"phone","relays":[],"v":1}"""))
        }
        assertFailsWith<IllegalArgumentException> {
            PhoneInvite.parse(encode("h",
                """{"kind":"phone","relays":[{"ws_url":"wss://x"}],"v":1}"""))
        }
    }
}

File: clients/rn/__tests__/network/phoneInvite.test.ts
/**
 * wyrdphone:// invite parsing.
 * Payload fixtures mirror exactly what registration.py mint_phone_invite
 * emits (sorted-key compact JSON, base64url, no padding).
 */
import {isPhoneInviteUrl, parsePhoneInvite} from '../../src/network/phoneInvite';

function encodeInvite(host: string, payload: object): string {
  const json = JSON.stringify(payload);
  const b64 = Buffer.from(json, 'utf8')
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
  return `wyrdphone://${host}/${b64}`;
}

const SELF_SIGNED_PAYLOAD = {
  household_id: 'hh-9ce1d3b57ebf',
  kind: 'phone',
  minted_at: 1781225931,
  relays: [
    {
      ca_fp: 'E5:F0:A2:3E',
      fp: '63:24:CB:6B',
      nats_password: 'pw-secret',
      nats_user: 'relay_phone',
      ws_url: 'wss://127.0.0.1:4443',
    },
  ],
  v: 1,
  zone_id: 'unspecified',
};

describe('phoneInvite', () => {
  it('detects wyrdphone URLs case-insensitively with surrounding whitespace', () => {
    expect(isPhoneInviteUrl('  wyrdphone://x/abc ')).toBe(true);
    expect(isPhoneInviteUrl('WYRDPHONE://x/abc')).toBe(true);
    expect(isPhoneInviteUrl('wyrdrelay://x/abc')).toBe(false);
    expect(isPhoneInviteUrl('https://example.org')).toBe(false);
  });

  it('parses a self-signed relay invite end-to-end', () => {
    const invite = parsePhoneInvite(
      encodeInvite('127.0.0.1:4443', SELF_SIGNED_PAYLOAD),
    );
    expect(invite.relays).toHaveLength(1);
    const r = invite.relays[0];
    expect(r.wsUrl).toBe('wss://127.0.0.1:4443');
    expect(r.natsUser).toBe('relay_phone');
    expect(r.natsPassword).toBe('pw-secret');
    expect(r.fp).toBe('63:24:CB:6B');
    expect(r.caFp).toBe('E5:F0:A2:3E');
    expect(invite.householdId).toBe('hh-9ce1d3b57ebf');
    expect(invite.mintedAt).toBe(1781225931);
  });

  it('maps "unspecified" sentinel fields to undefined', () => {
    const invite = parsePhoneInvite(
      encodeInvite('relay.example.org', SELF_SIGNED_PAYLOAD),
    );
    expect(invite.zoneId).toBeUndefined();
  });

  it('parses an ACME invite (no pin material, default port)', () => {
    const invite = parsePhoneInvite(
      encodeInvite('relay.example.org', {
        kind: 'phone',
        relays: [
          {
            nats_password: 'pw',
            nats_user: 'relay_phone',
            ws_url: 'wss://relay.example.org',
          },
        ],
        v: 1,
      }),
    );
    expect(invite.relays[0].fp).toBeUndefined();
    expect(invite.relays[0].caFp).toBeUndefined();
  });

  it('preserves relay failover ordering', () => {
    const invite = parsePhoneInvite(
      encodeInvite('a', {
        kind: 'phone',
        relays: [
          {nats_password: 'p1', nats_user: 'u1', ws_url: 'wss://first'},
          {nats_password: 'p2', nats_user: 'u2', ws_url: 'wss://second'},
        ],
        v: 1,
      }),
    );
    expect(invite.relays.map(r => r.wsUrl)).toEqual(['wss://first', 'wss://second']);
  });

  it('rejects malformed input with readable errors', () => {
    expect(() => parsePhoneInvite('https://nope')).toThrow(/wyrdphone/);
    expect(() => parsePhoneInvite('wyrdphone://host-only')).toThrow(/payload/);
    expect(() => parsePhoneInvite('wyrdphone://h/!!notb64!!')).toThrow(/not valid/);
    expect(() =>
      parsePhoneInvite(encodeInvite('h', {kind: 'zone', relays: [], v: 1})),
    ).toThrow(/Not a phone invite/);
    expect(() =>
      parsePhoneInvite(encodeInvite('h', {kind: 'phone', relays: [], v: 1})),
    ).toThrow(/no relays/);
    expect(() =>
      parsePhoneInvite(
        encodeInvite('h', {kind: 'phone', relays: [{ws_url: 'wss://x'}], v: 1}),
      ),
    ).toThrow(/incomplete/);
  });
});
